### PR TITLE
Feat: refactor resources/data to use the new serialization/deserializ…

### PR DESCRIPTION
**Pull request by TomerHeber**
…ation logic

### Solution

Refactoring around template_resource.
Improvements in readResourceData to support slices.

This refactoring is for two separate issues:
#374  and #376

### CODE REVIEW

File: client/template.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ type Template struct {
 	TerragruntVersion    string           `json:"terragruntVersion,omitempty"`
 	IsDeleted            bool             `json:"isDeleted,omitempty"`
 	BitbucketClientKey   string           `json:"bitbucketClientKey"`
-	IsGitHubEnterprise   bool             `json:"isGitHubEnterprise"`
+	IsGithubEnterprise   bool             `json:"isGitHubEnterprise"`
```
Comment (TomerHeber) [inline client/template.go:61]: readResourceData and writeResourceData would have converted this to "is_git_hub_..."
This change converts it to "is_github_..." which is how it's defined in the schema.

File: env0/resource_template.go [hunk 2]
```diff
@@ -24,107 +24,34 @@ func resourceTemplate() *schema.Resource {
 	}
 }
 
-func templateCreatePayloadFromParameters(d *schema.ResourceData) (client.TemplateCreatePayload, diag.Diagnostics) {
```
Comment (TomerHeber) [inline env0/resource_template.go:27]: Most if it is now done automatically by readResourceData - lots of code removed now (:

File: env0/utils.go [hunk 3]
```diff
@@ -107,23 +106,9 @@ func readResourceData(i interface{}, d *schema.ResourceData) error {
 				return fmt.Errorf("internal error - unhandled field pointer kind %v", fieldType.Elem().Kind())
 			}
 		case reflect.Slice:
-			switch fieldType {
```
Comment (TomerHeber) [inline env0/utils.go:110]: This now has a generic solution. see below.

File: env0/utils.go [hunk 4]
```diff
@@ -134,6 +119,54 @@ func readResourceData(i interface{}, d *schema.ResourceData) error {
 	return nil
 }
 
+func readResourceDataSliceStructHelper(field reflect.Value, resource interface{}) error {
```
Comment (TomerHeber) [inline env0/utils.go:122]: This is already tested in:
TestReadResourceDataModule - for slice of structs.
TestReadResourceDataNotificationProjectAssignment - for slice of strings.

File: env0/utils.go [hunk 5]
```diff
@@ -134,6 +119,54 @@ func readResourceData(i interface{}, d *schema.ResourceData) error {
...
+
+		field := val.Field(i)
+		field.Set(reflect.ValueOf(fieldValue))
+	}
+
+	return nil
+}
+
+// Extracts a list of values from the resourcedata, and writes it to the struct field.
+func readResourceDataSlice(field reflect.Value, resources []interface{}) error {
```
Comment (TomerHeber) [inline env0/utils.go:145]: It's hard to read/understand reflection code.
Please ping me if anything is unclear.

Comment (TomerHeber) [reply env0/utils.go:145]: This is already tested in utils_test.
For slice of structs and slice of strings.

File: env0/utils.go (hunk 3 above)
Comment (TomerHeber) [reply env0/utils.go:110]: The generic solution now works for most structs.
In this case, I needed support for this:
```
type TemplateSshKey struct {
	Id   string `json:"id"`
	Name string `json:"name"`
}

```

File: env0/utils.go [hunk 6]
```diff
@@ -107,23 +106,9 @@ func readResourceData(i interface{}, d *schema.ResourceData) error {
 				return fmt.Errorf("internal error - unhandled field pointer kind %v", fieldType.Elem().Kind())
 			}
 		case reflect.Slice:
-			switch fieldType {
-			case reflect.TypeOf([]client.ModuleSshKey{}):
```
Comment (TomerHeber) [inline env0/utils.go:111]: and remove specific use-cases (the generic use-case covers them).

File: env0/resource_template.go (hunk 2 above)
Comment (yaronya) [reply env0/resource_template.go:27]: Awesome! 👏

**Review by yaronya — APPROVED** [review]

**Pull request merged**